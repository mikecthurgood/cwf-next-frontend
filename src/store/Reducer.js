const reducer = (state, action) => {
    switch (action.type) {
        
        case 'setFilterSelection':
            return { ...state, filterSelection: action.data}
        case 'setLoginError':
            return { ...state, loginError: action.data}
        case 'setLoginMenuVisible':
            return { ...state, loginMenuVisible: action.data}
        case 'setMobileMenuVisible':
            return { ...state, mobileMenuVisible: action.data}
        case 'setSearchBarVisibility':
            return { ...state, searchBarVisible: action.data}
        case 'setSearchFilter':
            return { ...state, searchFilter: action.data}
        case 'setScrollPosition':
            return { ...state, scrollPosition: action.data}
        case 'setSignUpSuccess':
            return { ...state, signUpSuccess: action.data}
        case 'setSingleWall':
            return { ...state, singleWall: action.data}
        case 'setSortInputVisibility':
            return { ...state, sortInputVisible: action.data}
        case 'setUser':
            return { ...state, user: action.data}
        case 'setPostCode':
            return { ...state, userPostCode: action.data}
        case 'setWalls':
            return { ...state, walls: action.data}
        case 'signOut' :
            localStorage.removeItem('token')
            localStorage.removeItem('userName')
            localStorage.removeItem('userId')
            return { 
                ...state,
                user: {username: null, userId: null, isAuth: false, token: ''},
                mobileMenuVisible: false,
                loginMenuVisible: false
            }
        default: 
            throw new Error('Invalid action type dispatched')
    }
}
export default reducer