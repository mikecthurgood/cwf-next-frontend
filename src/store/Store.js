import React, { useReducer } from  'react';
import reducer from './Reducer'

const initialState = {
    filterSelection: [],
    loginError: false,
    loginMenuVisible: false,
    mobileMenuVisible: false,
    searchBarVisible: false,
    scrollPosition: 0,
    signUpSuccess: false,
    singleWall: {},
    sortInputVisible: false,
    user: { username: null, userId: null, isAuth: false, token: '' },
    userPostCode: '',
    walls: [],
}

const State = React.createContext(initialState)
const Dispatch = React.createContext(null)

const Store = ({children}) => {
    
    const [state, dispatch] = useReducer(reducer, initialState)

    return (
        <Dispatch.Provider value={dispatch} >
            <State.Provider value={state} >
                {children}
            </State.Provider>
        </Dispatch.Provider>
    )
}


export { State, Dispatch, Store }

