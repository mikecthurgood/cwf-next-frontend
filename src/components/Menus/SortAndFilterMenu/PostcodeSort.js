import React, {useState, useContext, useEffect} from 'react';
import Styled from 'styled-components';
import { Dispatch, State } from '../../../store/Store';

const PostcodeSort = ({setFilterMenuVisibility}) => {

    const dispatch = useContext(Dispatch)
    const { userPostCode } = useContext(State)
    const [postCodeInput, setPostCodeInput] = useState(userPostCode)
    const [postCodeInputVisible, setPostCodeInputVisible] = useState(true)

    useEffect(() => {
        if (userPostCode) {
            setPostCodeInputVisible(false)
        }
    },[userPostCode])

    function handleSetPostCode (e) {
        e.preventDefault()
        setPostCodeInputVisible(false)
        let newPostcode
        const postcodeRegEx = /[A-Z]{1,2}[0-9]{1,2} ?[0-9][A-Z]{2}/i; 
        if (!postCodeInput) {
            newPostcode = ''
            localStorage.removeItem('userPostcode')
        }
        const validPostcode = postcodeRegEx.test(postCodeInput); 
        if (validPostcode) {
            newPostcode = postCodeInput
            localStorage.setItem('userPostcode', postCodeInput)
            setFilterMenuVisibility()
            dispatch({ type: 'setPostCode', data: newPostcode })
        }
    }

    function handlePostCodeChange (e) {
        setPostCodeInput(e.target.value)
    }

    function clearPostcode () {
        setPostCodeInput('')
        localStorage.removeItem('userPostcode')
        dispatch({ type: 'setPostCode', data: '' })
        setPostCodeInputVisible(true)
    }

    return (
        <PostCodeSortContainer>
            <label className='filter__menu-form-heading'>
                    <strong>Sort by distance</strong>
            </label>
                <>
                    <div className='change-post-code-container'>
                        {userPostCode && ( 
                            <p className='postcode-confirmation'>Sorted by distance from <strong>{userPostCode.toUpperCase()}</strong></p>
                        )}
                        {userPostCode && (
                            <span>
                                <p className='changePostcode' onClick={() => {
                                    setPostCodeInputVisible(!postCodeInputVisible)
                                }}>
                                    {postCodeInputVisible ? 'Cancel' : 'Change'}
                                </p>
                                {!postCodeInputVisible && (
                                    <p className='changePostcode' onClick={clearPostcode}>{`Clear`}</p>
                                )}
                            </span>
                        )}
                        <form className={`filter__menu-sort-form ${postCodeInputVisible ? 'visible' : ''}`}
                            onSubmit={handleSetPostCode} 
                            autoComplete="off"
                        >
                            <input
                                type="text" 
                                placeholder='Enter Postcode' 
                                name='searchInput' 
                                className={`filter__menu-sort-input`} 
                                value={postCodeInput}
                                onChange={handlePostCodeChange} 
                            />
                             <SubmitButton type="submit" value='Set' />
                        </form>
                    </div>
                </>
               
        </PostCodeSortContainer>
    )
}

export default PostcodeSort

const PostCodeSortContainer = Styled.div`
    .filter__menu-sort-form {
        display: none;
        margin-bottom: -45px;
        &.visible {
            display: flex;
        }
        .filter__menu-sort-input {
            position: relative;
            left: 10px;
            font-size: 16px;
            padding: 10px 2px 10px 10px;
            text-align: left;
            max-width: 200px;
            min-width: 180px;
            border: 1px solid black;
            border-radius: 3px;
            margin-top: 5px;
            transition: opacity 0.25s ease-out;
            font-family: 'Muli', sans-serif;
            color: #282c34;
            &:focus {
                outline: none;
            }
        }
    }
    .change-post-code-container {
        width: 70%;
        padding-left: 20px;
        .filter__menu-sort-input {
            left: 0;
        }
        span {
            display: flex;
            justify-content: space-between;
            p {
                margin: 10px 0 0 0;
                &:hover{
                    color: #ed5656;
                    cursor: pointer;
                }
            }
        }
        p {
            margin: 10px 0 0 0; 
        }
    }
`

const SubmitButton = Styled.input`
    position: relative;
    top: 4.5px;
    height: 42.5px;
    width: 44px;
    text-align: center;
    padding: 5px;
    border: 1px solid black;
    background-color: #ed5656;
    color: white;
    font-family: 'Muli', sans-serif;
    border-radius: 0 5px 5px 0;
    &:active, &:focus {
        outline: none;
    }
    &:hover {
        background-color: cf4b4b;
    }

`
                