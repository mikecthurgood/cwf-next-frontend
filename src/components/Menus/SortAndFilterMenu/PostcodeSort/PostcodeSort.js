import React, {useState, useContext, useEffect} from 'react';
import Styled from 'styled-components';
import { Dispatch, State } from '../../../../store/Store';

const PostcodeSort = ({setFilterMenuVisibility}) => {

    const dispatch = useContext(Dispatch)
    const { userPostcode } = useContext(State)
    const [postcodeInput, setPostcodeInput] = useState(userPostcode)
    const [postcodeInputVisible, setPostcodeInputVisible] = useState(true)

    useEffect(() => {
        if (userPostcode) {
            setPostcodeInputVisible(false)
        }
    },[userPostcode])

    function handleSetPostcode (e) {
        e.preventDefault()
        setPostcodeInputVisible(false)
        let newPostcode
        const postcodeRegEx = /[A-Z]{1,2}[0-9]{1,2} ?[0-9][A-Z]{2}/i; 
        if (!postcodeInput) {
            newPostcode = ''
            localStorage.removeItem('userPostcode')
            setFilterMenuVisibility()
            dispatch({ type: 'setPostcode', data: newPostcode })
        }
        const validPostcode = postcodeRegEx.test(postcodeInput); 
        if (validPostcode) {
            newPostcode = postcodeInput
            localStorage.setItem('userPostcode', postcodeInput)
            setFilterMenuVisibility()
            dispatch({ type: 'setPostcode', data: newPostcode })
        }
    }

    function handlePostcodeChange (e) {
        setPostcodeInput(e.target.value)
    }

    function clearPostcode () {
        setPostcodeInput('')
        localStorage.removeItem('userPostcode')
        dispatch({ type: 'setPostcode', data: '' })
        setPostcodeInputVisible(true)
    }

    return (
        <PostcodeSortContainer>
            <label className='filter__menu-form-heading'>
                    <strong>Sort by distance</strong>
            </label>
                <>
                    <div className='change-post-code-container'>
                        {userPostcode && ( 
                            <p className='postcode-confirmation'>Sorted by distance from <strong>{userPostcode.toUpperCase()}</strong></p>
                        )}
                        {userPostcode && (
                            <span>
                                <p className='changePostcode' onClick={() => {
                                    setPostcodeInputVisible(!postcodeInputVisible)
                                }}>
                                    {postcodeInputVisible ? 'Cancel' : 'Change'}
                                </p>
                                {!postcodeInputVisible && (
                                    <p className='changePostcode' onClick={clearPostcode}>{`Clear`}</p>
                                )}
                            </span>
                        )}
                        <form className={`filter__menu-sort-form ${postcodeInputVisible ? 'visible' : ''}`}
                            onSubmit={handleSetPostcode} 
                            autoComplete="off"
                        >
                            <input
                                type="text" 
                                placeholder='Enter Postcode' 
                                name='searchInput' 
                                className={`filter__menu-sort-input`} 
                                value={postcodeInput}
                                onChange={handlePostcodeChange} 
                            />
                             <SubmitButton type="submit" value='Set' />
                        </form>
                    </div>
                </>
               
        </PostcodeSortContainer>
    )
}

export default PostcodeSort

const PostcodeSortContainer = Styled.div`
    .filter__menu-sort-form {
        display: none;
        margin-bottom: -45px;
        &.visible {
            display: flex;
        }
        .filter__menu-sort-input {
            position: relative;
            left: 10px;
            font-size: 16px;
            padding: 10px 2px 10px 10px;
            text-align: left;
            max-width: 200px;
            min-width: 180px;
            border: 1px solid black;
            border-radius: 3px;
            margin-top: 5px;
            transition: opacity 0.25s ease-out;
            font-family: 'Muli', sans-serif;
            color: #282c34;
            &:focus {
                outline: none;
            }
        }
    }
    .change-post-code-container {
        width: 70%;
        padding-left: 20px;
        .filter__menu-sort-input {
            left: 0;
        }
        span {
            display: flex;
            justify-content: space-between;
            p {
                margin: 10px 0 0 0;
                &:hover{
                    color: #ed5656;
                    cursor: pointer;
                }
            }
        }
        p {
            margin: 10px 0 0 0; 
        }
    }
`

const SubmitButton = Styled.input`
    position: relative;
    top: 4.5px;
    height: 42.5px;
    width: 44px;
    text-align: center;
    padding: 5px;
    border: 1px solid black;
    background-color: #ed5656;
    color: white;
    font-family: 'Muli', sans-serif;
    border-radius: 0 5px 5px 0;
    &:active, &:focus {
        outline: none;
    }
    &:hover {
        background-color: cf4b4b;
    }

`
                