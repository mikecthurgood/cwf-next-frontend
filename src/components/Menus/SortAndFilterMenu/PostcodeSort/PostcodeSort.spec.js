import React from 'react';
import { cleanup, fireEvent, render } from '@testing-library/react';
import { Dispatch, State } from '../../../../store/Store';
import '@testing-library/jest-dom';
import PostcodeSort from '.';


describe('Postcode Sort', () => {
    let dispatch;
    beforeEach(() => {
        dispatch = jest.fn();
    })

    afterEach(() => {
        cleanup();
    })
    const state = {
        userPostcode: '',
    };

    const setFilterMenuVisibility = jest.fn();

    it('should render the component', () => {
        const { container } = render(
            <Dispatch.Provider value={dispatch}>
                <State.Provider value={state}>
                    <PostcodeSort setFilterMenuVisibility={setFilterMenuVisibility} />
                </State.Provider>
            </Dispatch.Provider>
        );
        expect(container).toMatchSnapshot();
    });

    it('should display the user postcode if set and give option to change it', () => {
        state.userPostcode = 'te5 t1ng'
        const { container } = render(
            <Dispatch.Provider value={dispatch}>
                <State.Provider value={state}>
                    <PostcodeSort setFilterMenuVisibility={setFilterMenuVisibility} />
                </State.Provider>
            </Dispatch.Provider>
        );

        expect(container).toMatchSnapshot();
    })

    it('should open the postcode input if "change" button is clicked', () => {
        state.userPostcode = 'te5 t1ng'
        const { container, getByText } = render(
            <Dispatch.Provider value={dispatch}>
                <State.Provider value={state}>
                    <PostcodeSort setFilterMenuVisibility={setFilterMenuVisibility} />
                </State.Provider>
            </Dispatch.Provider>
        );

        const changePostcodeLink = getByText('Change');
        fireEvent.click(changePostcodeLink);
        expect(container).toMatchSnapshot();
    })

    it('should clear the postcode when "clear" button is clicked', () => {
        state.userPostcode = 'te5 t1ng'
        localStorage.setItem('userPostcode', 'te5 t1ng')

        const { container, getByText } = render(
            <Dispatch.Provider value={dispatch}>
                <State.Provider value={state}>
                    <PostcodeSort setFilterMenuVisibility={setFilterMenuVisibility} />
                </State.Provider>
            </Dispatch.Provider>
        );
        const clearPostcodeLink = getByText('Clear');
        fireEvent.click(clearPostcodeLink);

        expect(dispatch).toHaveBeenCalledWith({
            type: 'setPostcode',
            data: '',
        });
        expect(localStorage.getItem('userPostcode')).toBe(null);
        expect(container).toMatchSnapshot();
    })

    it('should change the postcode when a new postcode is entered', () => {
        state.userPostcode = 'te5 t1ng'
        localStorage.setItem('userPostcode', 'te5 t1ng')
        
        const { getByText, getByPlaceholderText } = render(
            <Dispatch.Provider value={dispatch}>
                <State.Provider value={state}>
                    <PostcodeSort setFilterMenuVisibility={setFilterMenuVisibility} />
                </State.Provider>
            </Dispatch.Provider>
        );
        const changePostcode = getByText('Change')
        fireEvent.click(changePostcode);

        const postcodeInput = getByPlaceholderText('Enter Postcode');
        fireEvent.change(postcodeInput, { target: { value: 'AA1 1AA' } })

        const submitPostcodeButton = getByText('Set');
        fireEvent.click(submitPostcodeButton);

        expect(dispatch).toHaveBeenCalledWith({
            type: 'setPostcode',
            data: 'AA1 1AA'
        })

        expect (setFilterMenuVisibility).toHaveBeenCalled();
    })

    it('should be able to clear postcode by submitting an empty string', () => {
        state.userPostcode = 'te5 t1ng'
        localStorage.setItem('userPostcode', 'te5 t1ng')
        
        const { getByText, getByPlaceholderText } = render(
            <Dispatch.Provider value={dispatch}>
                <State.Provider value={state}>
                    <PostcodeSort setFilterMenuVisibility={setFilterMenuVisibility} />
                </State.Provider>
            </Dispatch.Provider>
        );
        const changePostcode = getByText('Change')
        fireEvent.click(changePostcode);

        const postcodeInput = getByPlaceholderText('Enter Postcode');
        fireEvent.change(postcodeInput, { target: { value: '' } })

        const submitPostcodeButton = getByText('Set');
        fireEvent.click(submitPostcodeButton);

        expect(dispatch).toHaveBeenCalledWith({
            type: 'setPostcode',
            data: ''
        })

        expect (setFilterMenuVisibility).toHaveBeenCalled();
    })
})