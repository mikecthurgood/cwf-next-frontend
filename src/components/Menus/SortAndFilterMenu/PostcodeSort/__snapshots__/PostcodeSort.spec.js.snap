// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Postcode Sort should clear the postcode when "clear" button is clicked 1`] = `
<div>
  <div
    class="PostcodeSort__PostcodeSortContainer-kmt73j-0 fmYhrI"
  >
    <label
      class="filter__menu-form-heading"
    >
      <strong>
        Sort by distance
      </strong>
    </label>
    <div
      class="change-post-code-container"
    >
      <p
        class="postcode-confirmation"
      >
        Sorted by distance from 
        <strong>
          TE5 T1NG
        </strong>
      </p>
      <span>
        <p
          class="changePostcode"
        >
          Cancel
        </p>
      </span>
      <form
        autocomplete="off"
        class="filter__menu-sort-form visible"
      >
        <input
          class="filter__menu-sort-input"
          name="searchInput"
          placeholder="Enter Postcode"
          type="text"
          value=""
        />
        <input
          class="PostcodeSort__SubmitButton-kmt73j-1 bfOvYG"
          type="submit"
          value="Set"
        />
      </form>
    </div>
  </div>
</div>
`;

exports[`Postcode Sort should display the user postcode if set and give option to change it 1`] = `
<div>
  <div
    class="PostcodeSort__PostcodeSortContainer-kmt73j-0 fmYhrI"
  >
    <label
      class="filter__menu-form-heading"
    >
      <strong>
        Sort by distance
      </strong>
    </label>
    <div
      class="change-post-code-container"
    >
      <p
        class="postcode-confirmation"
      >
        Sorted by distance from 
        <strong>
          TE5 T1NG
        </strong>
      </p>
      <span>
        <p
          class="changePostcode"
        >
          Change
        </p>
        <p
          class="changePostcode"
        >
          Clear
        </p>
      </span>
      <form
        autocomplete="off"
        class="filter__menu-sort-form "
      >
        <input
          class="filter__menu-sort-input"
          name="searchInput"
          placeholder="Enter Postcode"
          type="text"
          value="te5 t1ng"
        />
        <input
          class="PostcodeSort__SubmitButton-kmt73j-1 bfOvYG"
          type="submit"
          value="Set"
        />
      </form>
    </div>
  </div>
</div>
`;

exports[`Postcode Sort should open the postcode input if "change" button is clicked 1`] = `
<div>
  <div
    class="PostcodeSort__PostcodeSortContainer-kmt73j-0 fmYhrI"
  >
    <label
      class="filter__menu-form-heading"
    >
      <strong>
        Sort by distance
      </strong>
    </label>
    <div
      class="change-post-code-container"
    >
      <p
        class="postcode-confirmation"
      >
        Sorted by distance from 
        <strong>
          TE5 T1NG
        </strong>
      </p>
      <span>
        <p
          class="changePostcode"
        >
          Cancel
        </p>
      </span>
      <form
        autocomplete="off"
        class="filter__menu-sort-form visible"
      >
        <input
          class="filter__menu-sort-input"
          name="searchInput"
          placeholder="Enter Postcode"
          type="text"
          value="te5 t1ng"
        />
        <input
          class="PostcodeSort__SubmitButton-kmt73j-1 bfOvYG"
          type="submit"
          value="Set"
        />
      </form>
    </div>
  </div>
</div>
`;

exports[`Postcode Sort should render the component 1`] = `
<div>
  <div
    class="PostcodeSort__PostcodeSortContainer-kmt73j-0 fmYhrI"
  >
    <label
      class="filter__menu-form-heading"
    >
      <strong>
        Sort by distance
      </strong>
    </label>
    <div
      class="change-post-code-container"
    >
      
      
      <form
        autocomplete="off"
        class="filter__menu-sort-form visible"
      >
        <input
          class="filter__menu-sort-input"
          name="searchInput"
          placeholder="Enter Postcode"
          type="text"
          value=""
        />
        <input
          class="PostcodeSort__SubmitButton-kmt73j-1 bfOvYG"
          type="submit"
          value="Set"
        />
      </form>
    </div>
  </div>
</div>
`;
